### PR TITLE
Network primitive

### LINKED ISSUE #19: Improve primitives for networking
The implementation for endpoints and adresses is incomplete. The resolver needs refactoring

### PR DESCRIPTION
**Pull request by emmett2020**
Hi, maikel, I have updated some network primitives. Mainly changes:
 1、check whether the address is a specific network type
 2、comparision operations for address/endpoint
 3、hash operations 
 4、unittests
 Thanks for your time.

### CODE REVIEW

File: source/sio/net/ip/address.hpp [hunk 1]
```diff
@@ -89,61 +95,138 @@ namespace sio::net::ip {
...
 
     /// Obtain an address object that represents any address.
     static address_v6 any() noexcept {
       return address_v6();
     }
 
     /// Obtain an address object that represents the loopback address.
-    static address_v6 loopback() noexcept;
+    static address_v6 loopback() noexcept {
+      address_v6 __any{};
```
Comment (maikel) [inline source/sio/net/ip/address.hpp]: I copied this file originally from a private branch of stdexec where identifiers needed to be uglified to target a standard library implementation. Here we don't need to prefix with `__` and it is even forbidden by the standard. So lets rename all `__foo` to `foo` and similarily all `_Bar` to `Bar`

File: source/sio/net/ip/address.hpp [hunk 2]
```diff
@@ -89,61 +95,138 @@ namespace sio::net::ip {
...
-    bool is_multicast_site_local() const noexcept;
+    std::basic_string<char, std::char_traits<char>, _Allocator>
+      to_string(_Allocator __alloc = _Allocator()) const {
+      char __buffer[INET6_ADDRSTRLEN + 1] = {};
+      ::inet_ntop(AF_INET6, &__addr_, __buffer, INET6_ADDRSTRLEN);
+      return std::string(__buffer, __alloc);
+    }
+
+    bool is_loopback() const noexcept {
+      return (__addr_.s6_addr[0] == 0) && (__addr_.s6_addr[1] == 0) && (__addr_.s6_addr[2] == 0)
```
Comment (maikel) [inline source/sio/net/ip/address.hpp]: Do you use exclusively `s6_addr` to not change the active member of the union?

**Review by maikel — COMMENTED** [review]
Just a minor nitpick, this is awesome

File: source/sio/net/ip/address.hpp (hunk 1 above)
Comment (emmett2020) [reply source/sio/net/ip/address.hpp]: yes, thanks.

**Pull request merged**